js/classes/entities/midground/enemies/flying_enemy_base.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * Class that handles functionality for flying enemies that move randomly along a 
+ * preset path. 
+ * 
+ * @author Trae Claar
+ */
+class FlyingEnemyBase {
+
+    /**
+     * Constructor for a FlyingEnemyBase.
+     * 
+     * @param {Entity} enemy the enemy associated with this EnemyBase
+     * @param {Vector} pos the initial position of the enemy
+     * @param {number} scaledSize the size of the enemy
+     * @param {number} speed the speed of the enemy
+     * @param {number} health the initial health of the enemy
+     * @param {function} onDeath callback function invoked when the enemy dies
+     * @param {Vector[]} path array of ordered Vectors that delineate the enemy's path
+     */
+    constructor(enemy, pos, scaledSize, speed, health, onDeath, path) {
+        this.enemy = enemy;
+        this.base = new EnemyBase(enemy, pos, scaledSize, speed, health, onDeath);
+
+        this.path = path;
+    
+        enemy.getDirection = () => this.getDirection();
+        
+        // set the enemy's target position as the nearest path node
+        for (let i = 0; i < path.length; i++) {
+            if (!this.target || Vector.distance(enemy.pos, path[i]) 
+                < Vector.distance(enemy.pos, this.target)) {
+                
+                this.target = i;
+            }
+        }
+
+        if (!enemy.takeDamage) {
+            enemy.takeDamage = (amount) => this.takeDamage(amount);
+        }
+    }
+
+    /**
+     * Get the direction in which the enemy is currently facing.
+     * 
+     * @returns {number} -1 for left and 1 for right
+     */
+    getDirection() {
+        return this.path[this.target].x - this.enemy.getCenter().x > 0 ? 1 : -1;
+    }
+
+    /**
+     * Does nothing. For compatibility with StatusEffect.
+     */
+    flee() {
+        
+    }
+
+    /**
+     * Update the enemy's position and target.
+     */
+    update() {
+        this.base.updateBoundingBox();
+
+        // if we've reached the target position, select a new target position
+        if (Vector.distance(this.enemy.pos, this.path[this.target]) < this.enemy.scaledSize.x / 2) {
+            const hasPrev = this.target > 0;
+            const hasNext = this.target < this.path.length - 1
+            
+            // if there are previous AND next nodes in the path, randomly choose between them
+            // otherwise, choose whichever is available
+            if (hasPrev && hasNext) {
+                this.target += ((Math.random() > 0.5) ? -1 : 1);
+            } else if (hasPrev) {
+                this.target -= 1;
+            } else if (hasNext) {
+                this.target += 1;
+            }
+        }
+
+        const velocity = Vector.multiply(Vector.unit(Vector.subtract(
```

Comment:
I'm a little confused on how velocity is found here. If I recall a unit vector is a value that represents a directional value such as velocity. So we are calculating the unit vector of the enemy's current position and where the enemy should go? Then multiplying velocity by that unit vector so that the enemy knows exactly where and how fast to go creating a b-line from start to target points?

Reply:
Close. We are multiplying the unit vector of the difference between the current position and target position, which gives us the direction from the start position to the target position, with the speed, which gives us the velocity ( denoting speed AND direction).

Reply:
Wow, super cool.